### PR TITLE
Fix expandable rows resetting columns

### PR DESCRIPTION
**Pull request by jake-bassett**
## Description
The change to use columnConfig$ as an Observable in table widget renderer caused this. The async pipe passed to columnConfig input on table was causing change detection. Fixed by comparing the values of the array instead of the array reference itself to determine if columnConfigs changed and we should initalizeColumns()

### Testing
Tested against each of the table modes

### CODE REVIEW

File: projects/components/src/table/table.component.ts [hunk 1]
```diff
@@ -358,8 +357,11 @@ export class TableComponent
       this.isTableFullPage = this.display === TableStyle.FullPage;
     }
 
-    if (changes.mode || changes.columnConfigs || changes.detailContent || changes.metadata) {
+    if (changes.mode || changes.detailContent || changes.metadata) {
       this.initializeColumns();
+    } else if (changes.columnConfigs) {
+      !isEqualIgnoreFunctions(changes.columnConfigs.previousValue, changes.columnConfigs.currentValue) &&
```
Comment (aaron-steinfeld) [inline projects/components/src/table/table.component.ts:363]: Why is this needed?

Comment (jake-bassett) [reply projects/components/src/table/table.component.ts:363]: Which part?

Comment (aaron-steinfeld) [reply projects/components/src/table/table.component.ts:363]: we're special casing column configs to not use reference equality - that's generally a responsibility delegated to the component passing in an input.

File: projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts [hunk 2]
```diff
@@ -129,7 +138,9 @@ export class TableWidgetRendererComponent
   }
 
   private buildColumns(): void {
-    this.columnConfigs$ = this.getColumnConfigs();
+    this.columnConfigs$ = this.getColumnConfigs().pipe(
```
Comment (anandtiwary) [inline projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts:141]: This would still call `getColumnConfigs` and its model counterparts twice for every change.

Comment (anandtiwary) [reply projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts:141]: `
diff --git a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
index 870d9ca..3a661b2 100644
--- a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
+++ b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
@@ -1,4 +1,5 @@
 import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Inject, OnInit } from '@angular/core';
+import { isEqualIgnoreFunctions } from '@hypertrace/common';
 import {
   FilterAttribute,
   FilterOperator,
@@ -17,7 +18,7 @@ import { Renderer } from '@hypertrace/hyperdash';
 import { RendererApi, RENDERER_API } from '@hypertrace/hyperdash-angular';
 import { capitalize, isEmpty } from 'lodash-es';
 import { BehaviorSubject, combineLatest, Observable, of, Subject } from 'rxjs';
-import { map, startWith, switchMap } from 'rxjs/operators';
+import { map, startWith, switchMap, pairwise, filter, share } from 'rxjs/operators';
 import { AttributeMetadata, toFilterAttributeType } from '../../../graphql/model/metadata/attribute-metadata';
 import { MetadataService } from '../../../services/metadata/metadata.service';
 import { InteractionHandler } from '../../interaction/interaction-handler';
@@ -97,7 +98,7 @@ export class TableWidgetRendererComponent
     super.ngOnInit();
 
     this.metadata$ = this.getScopeAttributes();
-    this.buildColumns();
+    this.columnConfigs$ = this.getColumnConfigs();
 
     this.combinedFilters$ = combineLatest([this.toggleFilterSubject, this.searchFilterSubject]).pipe(
       map(([toggleFilters, searchFilters]) => [...toggleFilters, ...searchFilters])
@@ -114,10 +115,6 @@ export class TableWidgetRendererComponent
     this.activeMode = this.model.mode;
   }
 
-  protected onModelChange(): void {
-    this.buildColumns();
-  }
-
   public getChildModel = (row: TableRow): object | undefined => this.model.getChildModel(row);
 
   protected fetchData(): Observable<TableDataSource<TableRow> | undefined> {
@@ -128,16 +125,19 @@ export class TableWidgetRendererComponent
     return this.model.style === TableStyle.FullPage;
   }
 
-  private buildColumns(): void {
-    this.columnConfigs$ = this.getColumnConfigs();
-  }
-
   private getScope(): Observable<string | undefined> {
     return this.data$!.pipe(map(data => data?.getScope()));
   }
 
   private getColumnConfigs(): Observable<TableColumnConfig[]> {
-    return this.getScope().pipe(switchMap(scope => this.model.getColumns(scope)));
+    return combineLatest([this.getScope(), this.api.change$.pipe(map(() => true),startWith(true))]).pipe(
+      switchMap(([scope]) => this.model.getColumns(scope)),
+      startWith([]),
+      pairwise(),
+      filter(([previous, current]) => !isEqualIgnoreFunctions(previous, current)),
+      map(([_, current]) => current),
+      share()
+    );
   }
 
   private getScopeAttributes(): Observable<FilterAttribute[]> {
diff --git a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
index 9524e08..12ee9e6 100644
--- a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
+++ b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
@@ -200,7 +200,7 @@ export class TableWidgetModel {
 
   public getChildModel(row: TableRow): object | undefined {
     if (this.mode === TableMode.Detail && this.childTemplate) {
-      const childModel = this.api.createChild<object>(this.childTemplate, this);
+      const childModel = this.api.createChild<object>(this.childTemplate);
       this.api.setVariable('row', row, childModel);
 
       return childModel;

`

Comment (anandtiwary) [reply projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts:141]:
```
diff --git a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
index 870d9ca..3a661b2 100644
--- a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
+++ b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
@@ -1,4 +1,5 @@
 import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Inject, OnInit } from '@angular/core';
+import { isEqualIgnoreFunctions } from '@hypertrace/common';
 import {
   FilterAttribute,
   FilterOperator,
@@ -17,7 +18,7 @@ import { Renderer } from '@hypertrace/hyperdash';
 import { RendererApi, RENDERER_API } from '@hypertrace/hyperdash-angular';
 import { capitalize, isEmpty } from 'lodash-es';
 import { BehaviorSubject, combineLatest, Observable, of, Subject } from 'rxjs';
-import { map, startWith, switchMap } from 'rxjs/operators';
+import { map, startWith, switchMap, pairwise, filter, share } from 'rxjs/operators';
 import { AttributeMetadata, toFilterAttributeType } from '../../../graphql/model/metadata/attribute-metadata';
 import { MetadataService } from '../../../services/metadata/metadata.service';
 import { InteractionHandler } from '../../interaction/interaction-handler';
@@ -97,7 +98,7 @@ export class TableWidgetRendererComponent
     super.ngOnInit();
 
     this.metadata$ = this.getScopeAttributes();
-    this.buildColumns();
+    this.columnConfigs$ = this.getColumnConfigs();
 
     this.combinedFilters$ = combineLatest([this.toggleFilterSubject, this.searchFilterSubject]).pipe(
       map(([toggleFilters, searchFilters]) => [...toggleFilters, ...searchFilters])
@@ -114,10 +115,6 @@ export class TableWidgetRendererComponent
     this.activeMode = this.model.mode;
   }
 
-  protected onModelChange(): void {
-    this.buildColumns();
-  }
-
   public getChildModel = (row: TableRow): object | undefined => this.model.getChildModel(row);
 
   protected fetchData(): Observable<TableDataSource<TableRow> | undefined> {
@@ -128,16 +125,19 @@ export class TableWidgetRendererComponent
     return this.model.style === TableStyle.FullPage;
   }
 
-  private buildColumns(): void {
-    this.columnConfigs$ = this.getColumnConfigs();
-  }
-
   private getScope(): Observable<string | undefined> {
     return this.data$!.pipe(map(data => data?.getScope()));
   }
 
   private getColumnConfigs(): Observable<TableColumnConfig[]> {
-    return this.getScope().pipe(switchMap(scope => this.model.getColumns(scope)));
+    return combineLatest([this.getScope(), this.api.change$.pipe(map(() => true),startWith(true))]).pipe(
+      switchMap(([scope]) => this.model.getColumns(scope)),
+      startWith([]),
+      pairwise(),
+      filter(([previous, current]) => !isEqualIgnoreFunctions(previous, current)),
+      map(([_, current]) => current),
+      share()
+    );
   }
 
   private getScopeAttributes(): Observable<FilterAttribute[]> {
diff --git a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
index 9524e08..12ee9e6 100644
--- a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
+++ b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget.model.ts
@@ -200,7 +200,7 @@ export class TableWidgetModel {
 
   public getChildModel(row: TableRow): object | undefined {
     if (this.mode === TableMode.Detail && this.childTemplate) {
-      const childModel = this.api.createChild<object>(this.childTemplate, this);
+      const childModel = this.api.createChild<object>(this.childTemplate);
       this.api.setVariable('row', row, childModel);
 
       return childModel;

```

Comment (anandtiwary) [reply projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts:141]:
```
diff --git a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
index 870d9ca..3a661b2 100644
--- a/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
+++ b/projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts
@@ -1,4 +1,5 @@
 import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Inject, OnInit } from '@angular/core';
+import { isEqualIgnoreFunctions } from '@hypertrace/common';
 import {
   FilterAttribute,
   FilterOperator,
@@ -17,7 +18,7 @@ import { Renderer } from '@hypertrace/hyperdash';
 import { RendererApi, RENDERER_API } from '@hypertrace/hyperdash-angular';
 import { capitalize, isEmpty } from 'lodash-es';
 import { BehaviorSubject, combineLatest, Observable, of, Subject } from 'rxjs';
-import { map, startWith, switchMap } from 'rxjs/operators';
+import { map, startWith, switchMap, pairwise, filter, share } from 'rxjs/operators';
 import { AttributeMetadata, toFilterAttributeType } from '../../../graphql/model/metadata/attribute-metadata';
 import { MetadataService } from '../../../services/metadata/metadata.service';
 import { InteractionHandler } from '../../interaction/interaction-handler';
@@ -97,7 +98,7 @@ export class TableWidgetRendererComponent
     super.ngOnInit();
 
     this.metadata$ = this.getScopeAttributes();
-    this.buildColumns();
+    this.columnConfigs$ = this.getColumnConfigs();
 
     this.combinedFilters$ = combineLatest([this.toggleFilterSubject, this.searchFilterSubject]).pipe(
       map(([toggleFilters, searchFilters]) => [...toggleFilters, ...searchFilters])
@@ -114,10 +115,6 @@ export class TableWidgetRendererComponent
     this.activeMode = this.model.mode;
   }
 
-  protected onModelChange(): void {
-    this.buildColumns();
-  }
-
   public getChildModel = (row: TableRow): object | undefined => this.model.getChildModel(row);
 
   protected fetchData(): Observable<TableDataSource<TableRow> | undefined> {
@@ -128,16 +125,19 @@ export class TableWidgetRendererComponent
     return this.model.style === TableStyle.FullPage;
   }
 
-  private buildColumns(): void {
-    this.columnConfigs$ = this.getColumnConfigs();
-  }
-
   private getScope(): Observable<string | undefined> {
     return this.data$!.pipe(map(data => data?.getScope()));
   }
 
   private getColumnConfigs(): Observable<TableColumnConfig[]> {
-    return this.getScope().pipe(switchMap(scope => this.model.getColumns(scope)));
+    return combineLatest([this.getScope(), this.api.change$.pipe(map(() => true),startWith(true))]).pipe(
+      switchMap(([scope]) => this.model.getColumns(scope)),
+      startWith([]),
+      pairwise(),
+      filter(([previous, current]) => !isEqualIgnoreFunctions(previous, current)),
+      map(([_, current]) => current),
+      share()
+    );
   }
 
   private getScopeAttributes(): Observable<FilterAttribute[]> {

```

Comment (anandtiwary) [reply projects/distributed-tracing/src/shared/dashboard/widgets/table/table-widget-renderer.component.ts:141]: How about this?

**Review by anandtiwary — APPROVED** [review]
